### ISSUE TITLE
No executable found matching command "dotnet-watch"

### ISSUE THREAD

**Issue by dodyg** [aspnet/DotNetTools#147]
`dotnet watch` generates the following error "No executable found matching command "dotnet-watch""

`dotnet --version` 1.0.0-preview2-003121

project.json

```
{
    "version": "1.0.0-*",
    "buildOptions": {
        "emitEntryPoint": true
    },

    "dependencies": {
        "RawRabbit": "1.9.4-dotnetcore",
        "Microsoft.AspNetCore.Hosting" : "1.0.0-*"
    },

    "frameworks": {
        "net451": {
            "dependencies": {
                "Microsoft.AspNetCore.Server.Kestrel": "1.0.0"
            }
        }
    },
    "tools": {
        "Microsoft.Dotnet.Watcher.Tools": {
            "version": "1.0.0-*",
            "imports": "portable-net451+win8"
        }
    }
}
```

program.cs

```
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Http;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Hosting;

namespace Inspection.Service.Monitor
{
    public class Startup
    {
        public void Configure(IApplicationBuilder app)
        {
            app.Run(context =>
            {
                return context.Response.WriteAsync("Hello world");
            });
        }
    }


    public class Program
    {
        public static void Main(string[] args)
        {
              var host = new WebHostBuilder()
                .UseKestrel()
                .UseStartup<Startup>()
                .Build();

            host.Run();
        }
    }
}
```

**Comment by BrennanConroy** [comment]
I hope its `project.json` not `package.json`. Also, did you run `dotnet restore` first?

**Comment by dodyg** [comment]
It's ```project.json```. Yes, I did run ```dotnet restore```. I also deleted ```project.lock.json``` and rerun ```dotnet restore```. 

- ```dotnet build``` works
- ```dotnet run``` works

I check ```project.lock.json``` and found this
```
 "tools": {
    ".NETCoreApp,Version=v1.0": {
      "Microsoft.Dotnet.Watcher.Tools/1.0.0-preview2-final": {
        "type": "package",
        "dependencies": {
          "Microsoft.DotNet.Watcher.Core": "1.0.0-preview2-final",
          "Microsoft.Extensions.CommandLineUtils": "1.0.0",
          "Microsoft.Extensions.Logging": "1.0.0",
          "Microsoft.Extensions.Logging.Console": "1.0.0",
          "Microsoft.NETCore.App": "1.0.0",
          "System.Runtime.InteropServices.RuntimeInformation": "4.0.0"
        },
        "compile": {
          "lib/netcoreapp1.0/dotnet-watch.dll": {}
        },
        "runtime": {
          "lib/netcoreapp1.0/dotnet-watch.dll": {}
        }
      }
    }
  },
  "projectFileToolGroups": {
    ".NETCoreApp,Version=v1.0": [
      "Microsoft.Dotnet.Watcher.Tools >= 1.0.0-*"
    ]
  }
```

So it looks fine.

**Comment by dodyg** [comment]
I put the code here on [github](https://github.com/dodyg/practical-aspnetcore/tree/master/hello-world-with-reload). I run on Windows 10.

**Comment by dodyg** [comment]
Damn, I fixed it by changing

```
   "tools": {
        "Microsoft.Dotnet.Watcher.Tools": {
            "version": "1.0.0-*",
            "imports": "portable-net451+win8"
        }
    }
```

to 
```
  "tools": {
        "Microsoft.Dotnet.Watcher.Tools": {
            "version": "1.0.0-preview2-final",
            "imports": "portable-net451+win8"
        }
```

I have no idea why ```1.0.0-*``` would not work. 

So the final version is 

```
{
    "version": "1.0.0-*",
    "buildOptions": {
        "emitEntryPoint": true
    },

    "dependencies": {
        "Microsoft.AspNetCore.Hosting" : "1.0.0-*",
        "Microsoft.Extensions.DependencyInjection": "1.0.0-*"
    },

    "frameworks": {
        "net451": {
            "dependencies": {
                "Microsoft.AspNetCore.Server.Kestrel": "1.0.0"
            }
        }
    },
    "tools": {
        "Microsoft.Dotnet.Watcher.Tools": {
            "version": "1.0.0-preview2-final",
            "imports": "portable-net451+win8"
        }
    }
}
```

**Comment by muratg** [comment]
@dodyg glad you get it working!  Closing this one.

cc @rrelyea, @piotrpmsft re: dotnet restore question. Folks, it looks like tools restore requires exact versions. Is this by design?

**Comment by rrelyea** [comment]
I was not able to reproduce the problem. both with a new project, or with your repo.
I ran "nuget locals -clear all" before testing each scenario.
This would be a bug, but I think there is another issue, or was mistakenly diagnosed. Would love to know.
@joelverhagen

**Comment by joelverhagen** [comment]
@muratg, tools versions should NOT have to be exact. In fact, the version specified explicitly above (`1.0.0-preview2-final`) is actually interpreted as [a version range](http://nugettoolsdev.azurewebsites.net/3.5.0-beta2/parse-version-range?versionRange=1.0.0-preview2-final) in this context (both tools and dependencies). Additionally, both existing versions of that package on NuGet.org [should satisfy](http://nugettoolsdev.azurewebsites.net/3.5.0-beta2/find-best-version-match?versionRange=1.0.0-*&versions=1.0.0-preview1-final%0D%0A1.0.0-preview2-final) the `1.0.0-*` spec, with `1.0.0-preview2-final` being preferred.

**Comment by dodyg** [comment]
It's a weird problem. Here's [a screen recording of it](http://webmshare.com/XGZba)

**Comment by victorhurdugaci** [comment]
What NuGet feeds do you use?

**Comment by muratg** [comment]
Opening this back to investigate the restore issues. It's not a dotnet-watch issue, but I'd like to get to the bottom of this.

@dodyg In addition to @victorhurdugaci's  question... what version of the package do you get when you restore when you have `*` in ther version? It looks like restore is succeeding so you're likely getting a package.

**Comment by dodyg** [comment]
![nuget](https://cloud.githubusercontent.com/assets/1761249/17275701/3e4bf40e-5711-11e6-91ef-6dee3b745d49.png)

**Comment by dodyg** [comment]
- I cleared the folder
- Ran ```dotnet restore```
- The package was correctly downloaded
- The command did not work

![downloaded](https://cloud.githubusercontent.com/assets/1761249/17276004/a5feaefc-571b-11e6-875f-fc295ab6d324.png)

**Comment by joelverhagen** [comment]
@dodyg, I tried those steps with the same CLI version (`1.0.0-preview2-003121`) and sources. I also tried running the steps on a different drive than your `.nuget` folder (which seems to be the case for you). I could not reproduce the problem. Could you check the following:

### After restoring

(could you verify these steps for both `1.0.0-*` and `1.0.0-preview2-final`)

Does the `project.lock.json` file have a `"tools"` entry that looks like this?

```json
  "tools": {
    ".NETCoreApp,Version=v1.0": {
      "Microsoft.Dotnet.Watcher.Tools/1.0.0-preview2-final": {
        "type": "package",
        "dependencies": {
          "Microsoft.DotNet.Watcher.Core": "1.0.0-preview2-final",
          "Microsoft.Extensions.CommandLineUtils": "1.0.0",
          "Microsoft.Extensions.Logging": "1.0.0",
          "Microsoft.Extensions.Logging.Console": "1.0.0",
          "Microsoft.NETCore.App": "1.0.0",
          "System.Runtime.InteropServices.RuntimeInformation": "4.0.0"
        },
        "compile": {
          "lib/netcoreapp1.0/dotnet-watch.dll": {}
        },
        "runtime": {
          "lib/netcoreapp1.0/dotnet-watch.dll": {}
        }
      }
    }
  },
```

Does the following directory look like this?

````
> dir %userprofile%\.nuget\packages\.tools\Microsoft.Dotnet.Watcher.Tools\1.0.0-preview2-final\netcoreapp1.0

 Volume in drive C is OSDisk
 Volume Serial Number is B4D0-EEBB

 Directory of C:\Users\jver\.nuget\packages\.tools\Microsoft.Dotnet.Watcher.Tools\1.0.0-preview2-final\netcoreapp1.0

08/01/2016  11:49 AM    <DIR>          .
08/01/2016  11:49 AM    <DIR>          ..
08/01/2016  11:49 AM           105,160 Microsoft.Dotnet.Watcher.Tools.deps.json
08/01/2016  11:53 AM           331,716 project.lock.json
               2 File(s)        436,876 bytes
               2 Dir(s)  138,070,646,784 bytes free
```

Essentially, I'd like the verify that `dotnet restore` is doing what it is supposed to be doing.

/cc @brthor, do you have any thoughts on what could be happening here?

**Comment by piotrpMSFT** [comment]
Let's see if the file system & lock file are what we expect them to be. If so, and we're still seeing an issue, we can look into why the CLI is not finding the executable.

**Comment by dodyg** [comment]
@joelverhagen 
``` 1.0.0-* ```
```
 "tools": {
    ".NETCoreApp,Version=v1.0": {
      "Microsoft.Dotnet.Watcher.Tools/1.0.0-preview2-final": {
        "type": "package",
        "dependencies": {
          "Microsoft.DotNet.Watcher.Core": "1.0.0-preview2-final",
          "Microsoft.Extensions.CommandLineUtils": "1.0.0",
          "Microsoft.Extensions.Logging": "1.0.0",
          "Microsoft.Extensions.Logging.Console": "1.0.0",
          "Microsoft.NETCore.App": "1.0.0",
          "System.Runtime.InteropServices.RuntimeInformation": "4.0.0"
        },
        "compile": {
          "lib/netcoreapp1.0/dotnet-watch.dll": {}
        },
        "runtime": {
          "lib/netcoreapp1.0/dotnet-watch.dll": {}
        }
      }
    }
```

``` 1.0.0-preview2-final ```
```
"tools": {
    ".NETCoreApp,Version=v1.0": {
      "Microsoft.Dotnet.Watcher.Tools/1.0.0-preview2-final": {
        "type": "package",
        "dependencies": {
          "Microsoft.DotNet.Watcher.Core": "1.0.0-preview2-final",
          "Microsoft.Extensions.CommandLineUtils": "1.0.0",
          "Microsoft.Extensions.Logging": "1.0.0",
          "Microsoft.Extensions.Logging.Console": "1.0.0",
          "Microsoft.NETCore.App": "1.0.0",
          "System.Runtime.InteropServices.RuntimeInformation": "4.0.0"
        },
        "compile": {
          "lib/netcoreapp1.0/dotnet-watch.dll": {}
        },
        "runtime": {
          "lib/netcoreapp1.0/dotnet-watch.dll": {}
        }
      }
    }
```

**Comment by dodyg** [comment]
```

C:\Program Files\RabbitMQ Server\rabbitmq_server-3.6.3\sbin>dir %userprofile%\.nuget\packages\.tools\Microsoft.Dotnet.Watcher.Tools\1.0.0-preview2-final\netcoreapp1.0
 Volume in drive C has no label.
 Volume Serial Number is E2E3-F4B4

 Directory of C:\Users\Atom\.nuget\packages\.tools\Microsoft.Dotnet.Watcher.Tools\1.0.0-preview2-final\netcoreapp1.0

07/27/2016  05:52 PM    <DIR>          .
07/27/2016  05:52 PM    <DIR>          ..
07/27/2016  05:52 PM           105,160 Microsoft.DotNet.Watcher.Tools.deps.json
08/02/2016  11:40 PM           331,716 project.lock.json
               2 File(s)        436,876 bytes
               2 Dir(s)  279,970,938,880 bytes free
```

**Comment by piotrpMSFT** [comment]
With the lock file above, does it succeed or fail?

**Comment by dodyg** [comment]
When version is ``` 1.0.0-* ```, it fails. When version is ```1.0.0-preview2-final``` it works.

Here's a weird sequence

- ``` 1.0.0-* ```
- ```dotnet restore```
- ```dotnet watch``` fails 
- change version to ```1.0.0-preview2-final```
- ```dotnet watch``` works

**Comment by playmobilman** [comment]
Did not work for me, followed steps provided by @dodyg but it keeps me saying `No executable found matching command dotnet-watch`

**Comment by brthor** [comment]
I'm also unable to repro this locally, using the repo @dodyg posted.
@dodyg @playmobilman 

Can you post the output of `dotnet --info` ? 

What installation mechanism did you use to acquire the cli?

**Comment by playmobilman** [comment]
Silly me, I forgot to restore packages, now its working fine. :)

**Comment by dodyg** [comment]
@brthor 

```
h:\Repository\Qualcomm\MTN\DeltaReport>dotnet --info                                                                                                                                                           
.NET Command Line Tools (1.0.0-preview2-003121)                                                                                                                                                                
                                                                                                                                                                                                               
Product Information:                                                                                                                                                                                           
 Version:            1.0.0-preview2-003121                                                                                                                                                                     
 Commit SHA-1 hash:  1e9d529bc5                                                                                                                                                                                
                                                                                                                                                                                                               
Runtime Environment:                                                                                                                                                                                           
 OS Name:     Windows                                                                                                                                                                                          
 OS Version:  10.0.10586                                                                                                                                                                                       
 OS Platform: Windows                                                                                                                                                                                          
 RID:         win10-x64                                                                                                                                                                                        
                                                  
```

I got them via https://www.microsoft.com/net/core#windows

**Comment by mikehaas763** [comment]
I'm on mac and am having the exact same issue. However, specifying the version explicitly does not fix it for me

**Comment by shanselman** [comment]
I'm seeing this also on 1.0.1. I can't get dotnet-watch to work at all. Something is wrong.

**Comment by geocine** [comment]
not working on 1.0.1

**Comment by aleksasavic3** [comment]
I fixed it by moving the `"Microsoft.Dotnet.Watcher.Tools/1.0.0-preview2-final"` from `dependencies` to `tools` in project.json. It looks like NuGet is putting it under `dependencies`.

**Comment by spockmonster** [comment]
I am trying to work through the MVA "Intro to ASP.Net Core" with Scott Hanselman, and the first lesson includes using dotnet watch.  However for me dotnet watch fails silently, "dotnet watch -?" executes and terminates with no output. 

I've tried several permutations of configuration in project.json, such as:
"tools": {
      "Microsoft.DotNet.Watcher.Tools": "1.0.0-*"
    }
which intalls "preview3-final,  ie:
  C:\Users\MY_USERNAME\.nuget\packages\.tools\Microsoft.DotNet.Watcher.Tools\1.0.0-preview3-final

I tried explicit version, ie:
"tools": {
      "Microsoft.DotNet.Watcher.Tools": {
        "version": "1.0.0-preview3-final",
        "imports": "portable-net451+win8"
      }
    }

At the cmd prompt in the same dir that contains project.json, typing "dotnet watch run" or "dotnet watch" or "dotnet watch -?" all exhibit the same behavior, the shell returns to the command prompt after about 1 second, and there is no output whatsoever.  It doesn't remain running, it terminates immediately, and there is no help output, etc.

If I specifically use preview2-final, then it works as expected.  So, this works:

"tools": {
      "Microsoft.DotNet.Watcher.Tools": {
        "version": "1.0.0-preview2-final",
        "imports": "portable-net451+win8"
      }
    }

Now I can run "dotnet watch run" and observe it recompiles when I change source code.

**Comment by brianlowe** [comment]
After trying all these permutations and still getting `No executable found matching command "dotnet-watch"` I finally realized that I was issuing the command in the wrong folder. The message is misleading.

Moving to `example/sometests` where the test project json exists and running "dotnet watch test" works as expected. 

It seems strange that I can use "dotnet test sometests" to run the tests in a project in a "sometests" folder, but "dotnet watch test sometests" fails so spectacularly.

**Comment by natemcmaster** [comment]
> I finally realized that I was issuing the command in the wrong folder. 

This is an external limitation of how the .NET Core CLI I finds and launches project tools extensions such as dotnet-watch. Unfortunately, there is nothing dotnet-watch can do about this until we get something like https://github.com/NuGet/Home/issues/1605.

**Comment by brianlowe** [comment]
Ah, that clears it up for me, partially.
I get that `dotnet` is on the path so can be called from anywhere. 
It seems that even though 'watch' doesn't appear as a command exposed in `dotnet --help` output, dotnet happily accepts it as an argument and uses it to form a new command `dotnet-watch` regardless of whether or not the current folder has a 'project.json' with a 'tools' section referencing "Microsoft.DotNet.Watcher.Tools".
The same thing happens for `dotnet foo`.
Maybe it would be better to check the argument passed against the built-in commands (new, restore, build, publish, run, test, pack} and then any tools referenced in the local 'project.json'. A message along the lines of "Foo is not a command and there is no tool referenced in the current folder's project.json adding 'Foo' as an extension"

**Comment by johnkors** [comment]
The requirement of calling `dotnet` from the specific folder where project.json recides instead of from a root folder (which is very common in build scripts) - is this in any of the docs for the CLI or the tools? 

Could not find it anywhere else than inside this issue. Possibly a time saver for everyone if this feature is documented. Not very obvious behaviour imo.

**Comment by brianlowe** [comment]
I don't think it's documented anywhere, and I don't suppose it will be documented any time soon. I guess from the dotnet-watch developers' perspective it's out of scope. They can't possibly predict every mistake any user will make when trying to use their tool, so they document only their expected 'correct' usage and leave it to end users to determine how their actual usage differs from what's documented. 

Having said that, there are some obvious or common or expected mistakes which should be documented to help users get past the initial barriers, or which could be picked up at runtime to raise a helpful error message. I thnk this is one such case and should have a more helpful error message.

**Comment by SeanHogg** [comment]
If you are experiencing this issue with the latest tooling (aka .csproj). You must modify your csproj and add:
<ItemGroup>
    <DotNetCliToolReference Include="Microsoft.DotNet.Watcher.Tools" Version="1.0.0-msbuild2-final" />
  </ItemGroup>
</Project>
This will allow you to call:
> dotnet watch run

**Comment by natemcmaster** [comment]
We released the "1.0.0" version of dotnet-watch (finally). As @SeanHogg  correctly points out, this can be installed using the `DotNetCliToolReference` syntax in csproj. 

I'll also add that another common mistake is to run `dotnet watch` from the wrong directory. You must change directories `cd` into the folder containing your csproj file.

We've also updated the project README to include simpler installation instructions. See https://github.com/aspnet/DotNetTools#how-to-install

I'm closing this as this issue was originally opened for project.json tooling. The 1.0.0 version of dotnet-watch does not support project.json.
